crates/gen/src/element_type.rs

```diff
@@ -0,0 +1,165 @@
+
+    FnPtr,  // Full method sig
+    Object, // System.Object
+    SZArray,
+    MVar, // Generic parameter in a method definition, represented as unsigned integer
+    CModReqd(TypeDefOrRef), // Required modifier
+    CModOpt(TypeDefOrRef), // Optional modifier
+    Internal,
+
+    Modifier, // Or'd with the following element types
+    Sentinel, // Sentinel for vararg method signature,
+
+    Pinned,
+
+    Type,         // System.Type
+    TaggedObject, // Boxed object (in custom attributes)
+    Field,        // Custom attribute field
+    Property,     // Custom attribute property
+    Enum,         // Custom attribute enum
+}
+
+impl ElementType {
+    pub fn from_blob(blob: &mut Blob) -> ElementType {
+        let code = blob.read_unsigned();
+        match code {
+            0x01 => ElementType::Void,
+            0x02 => ElementType::Boolean,
+            0x03 => ElementType::Char,
+            0x04 => ElementType::I1,
+            0x05 => ElementType::U1,
+            0x06 => ElementType::I2,
+            0x07 => ElementType::U2,
+            0x08 => ElementType::I4,
+            0x09 => ElementType::U4,
+            0x0a => ElementType::I8,
+            0x0b => ElementType::U8,
+            0x0c => ElementType::R4,
+            0x0d => ElementType::R8,
+            0x0e => ElementType::String,
+
+            0x0f => ElementType::Ptr(TypeSig::from_blob(blob)),
+            0x10 => ElementType::ByRef(TypeSig::from_blob(blob)),
+            0x11 => {
+                ElementType::ValueType(TypeDefOrRef::decode(blob.read_unsigned(), blob.file_index))
+            }
+            0x12 => ElementType::Class(TypeDefOrRef::decode(blob.read_unsigned(), blob.file_index)),
+            0x13 => ElementType::Var,
+            0x14 => ElementType::Array,
+            0x15 => ElementType::GenericInst,
+            0x16 => ElementType::TypedByRef,
+
+            0x18 => ElementType::I,
+            0x19 => ElementType::U,
+
+            0x1b => ElementType::FnPtr, // TODO: Followed by full function signature
+            0x1c => ElementType::Object,
+            0x1d => ElementType::SZArray,
+            0x1e => ElementType::MVar,
+            0x1f => {
+                ElementType::CModReqd(TypeDefOrRef::decode(blob.read_unsigned(), blob.file_index))
+            }
+            0x20 => {
+                ElementType::CModOpt(TypeDefOrRef::decode(blob.read_unsigned(), blob.file_index))
+            }
+            0x21 => ElementType::Internal,
+
+            0x40 => ElementType::Modifier,
+            0x41 => ElementType::Sentinel,
+
+            0x45 => ElementType::Pinned,
+
+            0x50 => ElementType::Type,
+            0x51 => ElementType::TaggedObject,
+            0x53 => ElementType::Field,
+            0x54 => ElementType::Property,
+            0x55 => ElementType::Enum,
+
+            unknown_type => panic!(format!("Unexpected ElementType: {:x}", unknown_type)),
+        }
+    }
+}
+
+impl std::fmt::Debug for ElementType {
```

Comment:
This implementation could be better. Enum needed this to have a Debug implementation so it could derive, but making this able to Derive took me down a rabbit hole of changing too many files so that's a future exercise.

Reply:
`Debug` was important when I was bootstrapping the parser and had no tokenization. Now that it's up and running I don't know that we need to keep `Debug` implemented. I certainly don't want to manage a bunch of extra code if it can't be derived.

Reply:
I personally don't see too much of a burden maintaining this particular debug implementation. It shouldn't change often (or ever?), and I too have used debug implementations for debugging and just for understanding what's happening.